### ISSUE TITLE
Multisite Indexing

### ISSUE THREAD

**Issue by allan23** [pantheon-systems/solr-power#41]
How should indexing be handled in a multisite environment?
Should each site have their own admin page or just one main page for the network admin?

CC: @joshkoenig

**Comment by joshkoenig** [comment]
This is a good question.

I do not believe that having different configuration or operations (e.g. what post types to index, index reset, etc) for individual sites in a network is a good idea. Solr doesn't have this notion built into itself, and so we shouldn't present a false set of options. There's still only once `schema.xml` for all the content, no good way to get stats for a particular site. 

We should set the assumption explicitly that this will not work out well if you're abusing a multisite setup to run a bunch of divergent sites off one install. 

So, it's one admin interface for the network admin.

I do think we will need to have some configuration options, ideally with the ability to hard-code with a `CONSTANT`(ala the other core site network config settings).

I can see the following scenarios:

1. Searches are not at all segmented meaning search is "network-wide"
2. All searches are automatically scoped to the site from which they launched via a facet. No way to search "network-wide".
3. Sites in-network are available as an optional facet (possibly a good one to consider for #37), meaning a user can limit their search to a set of sites in the network if they choose.

**Comment by joshkoenig** [comment]
Let's consider this for the next release.

**Comment by ataylorme** [comment]
I think we should consider resolving the existing issues and feature requests first and tackle multisite compatibility after a version 1.0 release.

**Comment by taija** [comment]
We are moving several large wordpress multisite installations to Pantheon soon, and would love to be able to use Solr. Fingers crossed for multisite support soon!

**Comment by ataylorme** [comment]
@danielbachhuber this would be a good candidate for work either before or after debug bar support.

@taija can you expand on how you plan on using Solr Power in those multisites?
* Will you be using search (native of the widget), `WP_Query` support or both? Which features of these items are important?
* What type of content will you be indexing?
    - post types
    - meta data
    - taxonomy data

**Comment by taija** [comment]
@ataylorme - we would be using it as a drop-in replacement for WP search (native). 

We would like to index post types and taxonomy data, in that priority order.

**Comment by ataylorme** [comment]
@taija regarding:

> 1. Searches are not at all segmented meaning search is "network-wide"
> 2. All searches are automatically scoped to the site from which they launched via a facet. No way to search "network-wide".

As a user which would you expect to be the default?

**Comment by taija** [comment]
@ataylorme - for us, we have built the expectation that search happens network-wide, as this works well for the type of content we have.

Ideally, this should be user-configurable on the Network Admin dashboard, but for us network-wide works great.

**Comment by ataylorme** [comment]
> Ideally, this should be user-configurable on the Network Admin dashboard, but for us network-wide works great.

@taija agreed - just looking for input on what the default would be. Also if we have to roll out one or the other first then iterate for support of both is a future release it's good to know which is more useful.

**Comment by taija** [comment]
@ataylorme - from my perspective, network-wide search is the way to go.

1. It seems like, of the two, this may be easier to implement
2. This is functionality that WP Core is bad at

**Comment by danielbachhuber** [comment]
> As a user which would you expect to be the default?

I think it makes sense to keep the default behavior to that of `WP_Query`: query against the site's posts table.

We could reuse the existing `solr_integrate` query parameter to query against the network. However, there will be at least a couple challenges to be aware of:

1. `SolrPower_WP_Query` will need to perform some amount of `switch_to_blog()` to ensure posts are being pulled from the correct database. `switch_to_blog()` has some amount of performance overhead.
2. Posts will be parsed with the PHP for the current site, not with the PHP of the originating site. This has the potential of being problematic. For instance, if a post has a shortcode that is registered for site A but not site B, the shortcode won't be rendered when the post is displayed for results of site B. Similarly, because permalinks are dynamically created, a post from site A could have an incorrect permalink (that may or may not properly redirect) when displayed on site B.

**Comment by ataylorme** [comment]
> SolrPower_WP_Query will need to perform some amount of switch_to_blog() to ensure posts are being pulled from the correct database.

When `solr_integrate` is `true` `SolrPower_WP_Query` doesn't hit the database - it hits Solr directly so I don't thing getting content will be an issue. We'll probably want to add an item to the Solr index for site ID though. We can then use that when querying out to a specific site rather than the entire index.

I think the issue would be with `WP_Query` indexing. Perhaps with the batch indexing we can further segment between blogs on the network as well. Keep a key for which site is currently being indexed and run through the existing start/resume logic. Once site A finishes move on to site B and so on. So when you pause/resume it's for the current site. We would need a new method to reset the entire index and start over from the first site on the network.

> Posts will be parsed with the PHP for the current site, not with the PHP of the originating site. 

That's definitely an issue.

**Comment by kshaner** [comment]
With the 1.2.0 release is it possible to do a network wide search or does this only support indexing single sites?  

Looking through the commits I couldn't find a way to pass the blogid param to the `SolrPower_Api->master_query` method.  Setting blog_id inside the function to false allows the network wide search.

Is this something that is in a future release or am I missing a way to handle this?

**Comment by danielbachhuber** [comment]
> With the 1.2.0 release is it possible to do a network wide search or does this only support indexing single sites?

It's possible to index all sites on the network, but the plugin only supports searching each site individually because of the reasons communicated in https://github.com/pantheon-systems/solr-power/issues/41#issuecomment-282048749

> Is this something that is in a future release or am I missing a way to handle this?

It probably won't ever be officially supported in the plugin. If this is for a Pantheon site, you're welcome to reach out through your account dashboard and we'll talk you through the ins and outs of an implementation.

**Comment by kshaner** [comment]
It is for a pantheon site, but locally at the moment and as a proof of concept.  I've found some ways to do it by modifying the solarium object through the `solr_query` filter.  I'll work through it or use the account dashboard.

**Comment by danielbachhuber** [comment]
@kshaner Sounds good. Feel free to open a new issue in this repo if there are additional filters, etc. that it would be useful to have.
